### PR TITLE
Project 20 Auth Julia Ros & Ann-Sofi Jönsson

### PR DESCRIPTION
**Pull request by Ann-Sofi**
Link to my deployed project:

https://project-authorization-2021.netlify.app/

### CODE REVIEW

File: backend/server.js [hunk 1]
```diff
@@ -1,29 +1,104 @@
 import express from 'express'
-import bodyParser from 'body-parser'
 import cors from 'cors'
 import mongoose from 'mongoose'
+import crypto from 'crypto'
+import bcrypt from 'bcrypt'
+import dotenv from 'dotenv'
+
+dotenv.config()
```
Comment (rrpersson) [inline backend/server.js:8]: Good use of dot env

File: backend/server.js [hunk 2]
```diff
@@ -1,29 +1,104 @@
...
 
-// Start defining your routes here
 app.get('/', (req, res) => {
   res.send('Hello world')
 })
 
-// Start the server
+app.get('/secrets', authenticateUser)
+app.get('/secrets', (req, res) => {
+  res.json({secret: 'This is a secret message'})
```
Comment (rrpersson) [inline backend/server.js:56]: Is it rellay secret if I can read it! ;)

File: frontend/src/components/SecretPage.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  transition: all 0.3s ease 0s;
+
+  :hover {
+    background: linear-gradient(90deg, #FC00FF 0%, #006cde 100%);
+  }
+
+  @media (min-width: 767px){
+    margin: 38px 0 35px;
+  }
+`
```
Comment (rrpersson) [inline frontend/src/components/SecretPage.js:115]: Nice use of styled components!

File: frontend/src/components/SignIn.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+          type='text'
+          value={username}
+          onChange={(e) => setUsername(e.target.value)}
+          placeholder='username'
+        />
+        <Input
+          type='password'
+          value={password}
+          onChange={(e) => setPassword(e.target.value)}
+          placeholder='•••••'
```
Comment (rrpersson) [inline frontend/src/components/SignIn.js:68]: Nice placeholder! :)

File: frontend/src/reusables/urls.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+const BASE_URL = 'https://julia-annsofi-secret-api.herokuapp.com'
+
+export const API_URL = (slug) => `${BASE_URL}/${slug}`
```
Comment (rrpersson) [inline frontend/src/reusables/urls.js:3]: Good use of an urls file

**Review by rrpersson — COMMENTED** [review]
Well done on the project. I like your styling and color scheme! 
Good use of styled components as well!
Your code is easy to follow and it looks easy to maintain as well!

File: README.md [hunk 6]
```diff
@@ -1,13 +1,38 @@
...
+* Styled components
+
+
+## View it live ✨
+
+Finished project:
+https://project-authorization-2021.netlify.app/
+
+Backend:
+https://julia-annsofi-secret-api.herokuapp.com/
```
Comment (CarolineF3) [inline README.md:38]: Really nice ReadMe, very neat and descriptive. :)

File: README.md [hunk 7]
```diff
@@ -1,13 +1,38 @@
...
+* React
+* Redux
+* React router
+* Styled components
+
+
+## View it live ✨
+
+Finished project:
+https://project-authorization-2021.netlify.app/
```
Comment (KarolCislowski) [inline README.md:35]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```

image

File: backend/server.js [hunk 8]
```diff
@@ -1,29 +1,104 @@
...
+  },
+  password: {
+    type: String,
+    required: true
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex')
+  }
+})
```
Comment (KarolCislowski) [inline backend/server.js:28]: nice model 👍🏻

File: backend/server.js [hunk 9]
```diff
@@ -1,29 +1,104 @@
...
+        username: user.username,
+        accessToken: user.accessToken
+      })
+    } else {
+      res.status(404).json({ success: false, message: 'User or password not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ success: false, message: 'Invalid request', error })
+  }
+})
```
Comment (KarolCislowski) [inline backend/server.js:100]: nice middlewares 👍🏻

File: frontend/src/components/SecretPage.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+  }, [accessToken, history])
+
+
+  const handleClick = () => {
+    batch(() => {
+      dispatch(user.actions.setAccessToken(null))
+      dispatch(user.actions.setUsername(null))
+    })
```
Comment (KarolCislowski) [inline frontend/src/components/SecretPage.js:26]: Nice 👍🏻

File: frontend/src/components/SignIn.js [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({ username, password })
+    }
+
+    fetch(API_URL(status), options)
+      .then(res => res.json())
+      .then(data => {
+        console.log(data)
```
Comment (KarolCislowski) [inline frontend/src/components/SignIn.js:40]: 🧹 console.log 🧹

File: frontend/src/reducers/user.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    setAccessToken: (store, action) => {
+      store.accessToken = action.payload
+    },
+    setErrors: (store, action) => {
+      store.errors = action.payload
+    }
+  }
+})
+
+export default user
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:23]: You need to store your data in the local storage to avoid logging out your user after refreshing the page.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It is really nice app but you need to clean it up and persist your user in the local storage to avoid refreshing.

Because your graduation is so soon please let me know on slack (a short note with a link to this PR) when you finish fixing your app.

**Review by KarolCislowski — APPROVED** [review]
Nice project, you just need to fix the netlify issue by adding `_redirects` file to the public folder.

**Pull request closed without merging**